### PR TITLE
closes #16

### LINKED ISSUE #16: split preprocessor-related classes to different file
with the current file structure, it is not possible to implement the **preprocessor** ```#if``` statement, as it would overwrite (or be overwritten by) the **C** ```if ()``` statement.

I would propose to update the file structure to:
```
cgen
+-- cgen
    +-- __init__.py
    +-- cuda.py
    +-- ispc.py
    +-- mapper.py
    +-- opencl.py
    +-- preprocessor
        +-- __init__.py
```
and implement the preprocessor ```#if``` (and others preprocessor statements?) in the file ```cgen/preprocessor/__init__.py```.
this would provide access to the preprocessor functionalities like this:

```python
import cgen as c
import cgen.preprocessor as pp
func = c.FunctionBody(
    c.FunctionDeclaration(c.Const(c.Pointer(c.Value("char", "greet"))), []),
    c.Block([
        pp.If('(a == 0)',
            [
                c.If('b == 0', c.Block([
                    c.Statement('return "hello world"')
                ]))
            ],
            [])
        ])
    )

print(func)
```
and will output something like:
```
char const *greet()
{
  #if (a == 0)
  if (b == 0)
  {
    return "hello world";
  }
  #endif
}
```

### PR DESCRIPTION
**Pull request by Sauci**
the following classes should now be imported from cgen.preprocessor:
Define
Include
Pragma
IfDef
IfNDef
if imported from cgen, a deprecation warning will be issued.

### CODE REVIEW

File: cgen/__init__.py [hunk 1]
```diff
@@ -26,6 +26,18 @@
 
 import numpy
 from pytools import memoize_method, memoize
+import warnings
+
+from cgen.preprocessor import Generable, Block, Collection, Comment, Line
```
Comment (inducer) [inline cgen/__init__.py]: These don't make sense in that submodule.

You could create an "infrastructure" submodule perhaps.

File: cgen/__init__.py [hunk 2]
```diff
@@ -26,6 +26,18 @@
...
+from cgen.preprocessor import Define as _Define
+from cgen.preprocessor import Include as _Include
+from cgen.preprocessor import Pragma as _Pragma
+from cgen.preprocessor import IfDef as _IfDef
+from cgen.preprocessor import IfNDef as _IfNDef
```
Comment (inducer) [inline cgen/__init__.py]: How come this is not a single, comma-separated import?

File: setup.py [hunk 3]
```diff
@@ -29,7 +29,10 @@
         license="MIT",
         url="http://documen.tician.de/cgen/",
 
-        packages=["cgen"],
+        packages=[
+            "cgen",
+            "cgen.preprocessor"
+        ],
```
Comment (inducer) [inline setup.py]: This should just use `setuptools.find_packages()`

File: doc/source/cgen.rst [hunk 4]
```diff
@@ -43,6 +43,12 @@ Preprocessor Code
 .. autoclass:: Pragma
     :show-inheritance:
 
+.. autoclass:: IfDef
```
Comment (inducer) [inline doc/source/cgen.rst]: Could you reflect the submodule in the docs?

File: cgen/preprocessor/__init__.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
+class Generable(object):
```
Comment (inducer) [inline cgen/preprocessor/__init__.py]: How come this is a whole subfolder?

File: cgen/__init__.py (hunk 1 above)
Comment (Sauci) [reply cgen/__init__.py]: I agree, those classes should not be part of the preprocessor source file, I will move Generable, ```Block```, ```Collection``` and ```Line``` to an **infrastructure.py** file. I guess the ```Comment``` class should be part of the **preprocessor.py** file (at least not the new **infrastructure.py**).

File: cgen/__init__.py (hunk 2 above)
Comment (Sauci) [reply cgen/__init__.py]: Right, I will update the import statement as:
```python
from cgen.preprocessor import (
    Define as _Define,
    Include as _Include,
    Pragma as _Pragma,
    IfDef as _IfDef,
    IfNDef as _IfNDef)
```

File: cgen/__init__.py [hunk 6]
```diff
@@ -26,6 +26,18 @@
...
+
+from cgen.infrastructure import *
+from cgen.preprocessor import (Define as _Define,
+                               Include as _Include,
+                               Pragma as _Pragma,
+                               IfDef as _IfDef,
+                               IfNDef as _IfNDef)
+
+
+warnings.simplefilter('always', DeprecationWarning)
```
Comment (inducer) [inline cgen/__init__.py]: It is not the business of a submodule to configure warnings for a user.

File: cgen/preprocessor.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
+from cgen.infrastructure import Generable, Line, Module
+
+
```
Comment (inducer) [inline cgen/preprocessor.py]: Could you include a `__doc__ = ...` here, to work with the `automodule`?

File: cgen/preprocessor.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        self.filename = filename
+        self.system = system
+
+    def generate(self):
+        if self.system:
+            yield "#include <%s>" % self.filename
+        else:
+            yield "#include \"%s\"" % self.filename
+
+    mapper_method = "map_pp_include"
```
Comment (inducer) [inline cgen/preprocessor.py]: While I like the consistency, renaming this will break compatibility.

File: cgen/preprocessor.py (hunk 7 above)
Comment (Sauci) [reply cgen/preprocessor.py]: you mean adding the infrastructure module in the ```doc/source/cgen.rst```?

**Comment by inducer** [comment]
There are merge conflicts--could you take a look?

**Pull request closed without merging**